---
title: "Variant-specific inflation factors: a tutorial"
author: "Tamar Sofer"
header-includes:
    - \usepackage{setspace}\doublespacing
date: "`r Sys.Date()`"
output: 
  pdf_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# About this tutorial 
This tutorial demonstrates how to use the R functions provided with the manuscript "Population Stratification at the Phenotypic Variance level and Implication for the Analysis of Whole Genome Sequencing Data from Multiple Studies" to investigate potential mis-calibration of test statistics when testing genetic variant associations with a quantitative trait. P-values may be miscalibrated when individual-level data from multiple studies, or when multiple race/ethnic groups are pooled together, and both allele frequencies and trait residual variances differ between the groups.

In what follows, we will use data that were simulated in advance to compute residual variances and variant allele frequencies, and to test variant associations. Then, we will use this information to generate a figure with multiple panels of QQ-plots, each demonstrating inflation patterns in a different set of genetic variants. 

Note: this tutorial uses genetic data saved on a GDS file, and therefore, uses a specific set of R functions to perform tasks such as association testing. You can use other tools and skip to the point where you already have allele frequencies and p-values computed, even if they were computed in another software. 

# Preparing for analysis based on simulated data
In the following code we load packages, "source" functions used, and load the simulated data. 

Clean up the workspace, load required packages:
```{r}
rm(list = ls())
setwd("~/Documents/GitHub/Variant_specific_inflation")
source("variant_specific_inflation_functions.R")

require(GWASTools)
require(dummies)
require(data.table)
require(GENESIS)
```

Read genotype data, stored in a SNP-GDS file: 
```{r}
gds <- GdsGenotypeReader("Genotypes.gds")
scanID <- getScanID(gds)
snpID <- getSnpID(gds)
geno <- getGenotype(gds)
rownames(geno) <- snpID
colnames(geno) <- scanID
close(gds)
```

Load the phenotype data. The data was simulated and stored in a data.frame which we will call \texttt{phen}. The quantitative trait we will use is called \texttt{trait}. It the simulations, we assume that there are three different groups, \texttt{g1, g2, g3}, each has different error standard deviation.  Assuming that in genetic analysis we will only adjust for age, we regress the trait on age and take the residuals to compute residual standard deviation. 
```{r}
# load the simulated data:
phen <- getobj("Phenotypes.RData")
phen <- data.table(phen)

# this is how the data looks like:
phen
# compute residual standard deviations by group after regressing on age:
residual_SDs <- phen[, sd(lm(trait ~ age)$resid), by = "group"]
residual_SDs_vec <- residual_SDs$V1
names(residual_SDs_vec) <- residual_SDs$group

# these are the estimated residual standard deviations: 
residual_SDs_vec
```

The approximate inflation factors will also use the sample sizes of the different groups that were pooled together:

```{r}
# compute group sample sizes
ns <- table(phen$group) 

# make sure that the order matches that of the residual SDs vector:
ns <- ns[names(residual_SDs_vec)] 
```

We are going to perform association testing with the genotypes. We will now compute allele frequencies across the genotypes that will be tested. These frequencies will be used when computing inflation factors. In the following code, we compute allele frequencies by group. One could use built-in functions, but here is a code that assumes genotypes are on autosomes and computes the frequencies:

```{r}
freq_list <- vector(mode = "list", length = length(unique(phen$group)))
names(freq_list) <- unique(phen$group)
for (i in 1:length(freq_list)){
	freq_list[[i]] <- rowMeans(geno[,phen[group == names(freq_list[i])]$scanID], 
	                                                              na.rm = TRUE)/2
}

# put the frequencies in a data.frame:
eafs <- do.call(cbind, freq_list)

# make sure that the columns order matches the SD vector:
eafs <- eafs[,names(residual_SDs_vec)]
```

# Computing approximate inflation factors based on allele frequencies, sample sizes, and residual standard deviations
Here we use the function \texttt{compute\_variant\_inflation\_factor} which we provide.

```{r}
expected_inf <- rep(NA, nrow(eafs))
for (j in 1:length(expected_inf)){
		cur_eafs <- eafs[j,]
		if (sd(cur_eafs) == 0) next
		expected_inf[j] <- 
		  compute_variant_inflation_factor(eafs = cur_eafs, 
		                                   ns = ns, 
		                                   sigmas = residual_SDs_vec)[["Inflation_factor"]]
}

exp_inf <- data.frame(variant_id = snpID, 
                      expected_inf = expected_inf, 
                      stringsAsFactors = FALSE)
```

# Performing associating testing: homogeneous variance model
We use the R/Bioconductor GENESIS package. We first fit a "null model", and then use it for association testing. In the code, we fit the null model twice, because we use the fully-adjusted two stage procedure described in Sofer et al. (2019, Gen Epi).
```{r}
phen$group <- as.factor(phen$group)
phen <- data.frame(phen)
# fit null model (homogeneous residual variance).
nullmod_homogeneous <- fitNullModel(phen, 
                                    outcome = "trait", 
                                    covars = c("group","age"), 
                                    verbose = FALSE)
nullmod_homogeneous_rn <- nullModelInvNorm(nullmod_homogeneous, 
                                           norm.option = "all", 
                                           rescale = "residSD",
                                           verbose = FALSE)


### association tests
gds <- GdsGenotypeReader("Genotypes.gds")
genoData <- GenotypeData(gds, scanAnnot = ScanAnnotationDataFrame(phen))
iterator <- GenotypeBlockIterator(genoData)
assoc_homogeneous <- assocTestSingle(iterator, nullmod_homogeneous_rn)
close(gds)
```

# Making a QQ-plot figure by categories of inflation factors: only homogeneous variance model
We use the function \texttt{qq\_plot\_by\_region} to visualize inflation in sets of genetic variants. We use the individual inflation factors that we compued. To define the values for categorizing a variant as "inflated" we used 1.03 and higher; we set the value for which we categories a variant as "deflated" to 0.97 and lower. Finally, we provide a file name for the figure to print to. 

Note that you can include results from a few different models (i.e. more than two models) according to the number of columns in the R \texttt{data.frame} that contain p-values. 

```{r}
pval_df <- data.frame(variant_id = assoc_homogeneous$variant.id,	
						homogeneous_variance = assoc_homogeneous$Score.pval)
exp_inf <- exp_inf[match(pval_df$variant_id, exp_inf$variant_id),]

qq_plot_by_region(pval_df = pval_df[,c("homogeneous_variance"), drop = FALSE], 
					expected_inf = exp_inf$expected_inf, 
					inflated_val = 1.03, 
					deflated_val = 0.97, 
					thin_high_pvals = 1e5, 
					figure_file_name = "qq_plots_homo_var.png")
```
\includegraphics{qq_plots_homo_var.png}

# Performing associating testing: heterogeneos variance model
```{r}
nullmod_hetero <- fitNullModel(phen, 
                               outcome = "trait", 
                               covars = c("group","age"), 
                               group.var = "group", 
                               verbose = FALSE)
nullmod_hetero_rn <- nullModelInvNorm(nullmod_hetero, 
                                      norm.option = "all", 
                                      rescale = "residSD", 
                                      verbose = FALSE)

# testing again:
gds <- GdsGenotypeReader("Genotypes.gds")
genoData <- GenotypeData(gds, scanAnnot = ScanAnnotationDataFrame(phen))
iterator <- GenotypeBlockIterator(genoData)
assoc_hetero <- assocTestSingle(iterator, nullmod_hetero_rn)
close(gds)
```

# Making a QQ-plot figure by categories of inflation factors: both homogeneous and stratified variance models
```{r}

pval_df <- data.frame(variant_id = assoc_homogeneous$variant.id,	
						homogeneous_variance = assoc_homogeneous$Score.pval, 
							stratified_variance = assoc_hetero$Score.pval)

exp_inf <- exp_inf[match(pval_df$variant_id, exp_inf$variant_id),]

qq_plot_by_region(pval_df = pval_df[,c("homogeneous_variance","stratified_variance")], 
					expected_inf = exp_inf$expected_inf, 
					inflated_val = 1.03, 
					deflated_val = 0.97, 
					thin_high_pvals = 1e5, 
					figure_file_name = "qq_plots_homo_strat_var.png")

```
\includegraphics{qq_plots_homo_strat_var.png}

# Final note
The figures here may not be very impressive. This is because we used a very small dataset, with a small number of people and a small number of variants. For the stratified variance model, this is a small number of people to estimate the group-specific variances. This is likely what causes the perhaps deflation pattern seen in the stratified variance model in the "inflated" variants category. 


# Record R package versions
```{r}
sessionInfo()
```
